tests/validate/test_facades.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2023 Canonical Ltd.
+# Licensed under the Apache V2, see LICENCE file for details.
+
+import importlib
+import re
+from pathlib import Path
+from typing import Dict, List, Set, Tuple, TypedDict, cast
+
+import pytest
+
+from juju.client import connection, _definitions
+
+
+class Versions(TypedDict, total=True):
+    versions: List[int]
+
+
+ClientFacades = Dict[str, Versions]
+
+
+@pytest.fixture
+def project_root(pytestconfig: pytest.Config) -> Path:
+    return pytestconfig.rootpath
+
+
+def test_client_facades(project_root: Path) -> None:
+    client_facades = cast(ClientFacades, connection.client_facades)
+    good_facades = make_client_facades_from_generated_code(project_root)
+    assert {
+        k: v['versions'] for k, v in client_facades.items()
+    } == {
+        k: v['versions'] for k, v in good_facades.items()
+    }
+
+
+def make_client_facades_from_generated_code(project_root: Path) -> ClientFacades:
+    """Return a client_facades dictionary from generated code under project_root.
+    """
+    files_by_version: List[Tuple[int, Path]] = []
+    # [(facade_version, Path), ...]
+    for file in (project_root / 'juju' / 'client').glob('_client[0-9]*.py'):
+        files_by_version.append((_version_from_filename(file), file))
+    files_by_version.sort()
+
+    # _clientN.py files import * from _definitions
+    # so we will ignore any names from there
+    ignore = dir(_definitions)
+
+    facades_by_version: Dict[int, Set[str]] = {}
+    # {facade_version: {facade_name, ...}, ...}
+    for version, file in files_by_version:
+        module = importlib.import_module(f'juju.client.{file.stem}')
+        facades = {
+            name.removesuffix("Facade")
+            for name in dir(module)
+            if not (name.startswith('_') or name in ignore)
+        }
+        facades_by_version[version] = facades
```

Comment:
Would it be better to both distinguish classes that are facades and get it's versions here?

```py
  class MachinerFacade(Type):
      name = 'Machiner'
      version = 5
```

the .version attr?

Reply:
True, here I'm just assuming that the version in the filename is accurate. It would make sense to check that. We could check that the name matches as well. I'm guessing they have to match because of how codgen works, but we could add tests for this to make it easier to debug if something changed with that and breaks what we're trying to test here.

I wonder if you're also asking about building the `client_facades` mapping from facade name to version numbers in this pass instead of this intermediate mapping from version numbers to facades. I think you're right here too, this would be much more direct.

Reply:
Sorry I didn't mean to suggest extra checks.

I was proposing to rely on codegen being correct and to simplify stuff, in effect to remove checks.

Reply:
High level comment:

It's pretty common to write complex code and extra tests while developing something, because we're dipping into the unknown and learning how it actually works. It's fine to have fancy stuff in the branch history.

When the PR is getting ready to be merged, I would normally expect that things get simplified / streamlined and/or more understandable option chosen. I'd rather merge polished, concise code.

Reply:
That makes sense, thank you